mlir/include/mlir/Dialect/Rock/IR/RockOps.td

```diff
@@ -168,6 +174,30 @@ def Rock_GemmOp :
   }];
 }
 
+
+def Rock_ReduceOp :
+  Rock_Op<"reduce", [AllElementTypesMatch<["in", "out"]>]>,
+  Arguments<(ins
+    Arg<TensorOrMemRefOf<[F32]>, "input", [MemRead]>:$in,
+    Arg<TensorOrMemRefOf<[F32]>, "output", [MemRead, MemWrite]>:$out,
+    ReduceMethodAttr:$reduceMethod,
+    IndexAttr:$axis,
+    I32Attr:$blockSize,
+    I32Attr:$gridSize,
+    OptionalAttr<BoolAttr>:$useLDS,
+    OptionalAttr<BoolAttr>:$useDPP
+  )>,
+  Results<(outs Optional<TensorOf<[F32]>>:$result)> {
+  let summary = "Axes-wide reduction operation";
+  let hasVerifier = 1;
+  let assemblyFormat = [{
+    $reduceMethod `(` operands `)` attr-dict `:` `(` type($in) `,` type($out) `)` (`->` type($result)^)?
```

Comment:
For the sake of my sanity, can we give this a nicer assembly format?

Perhaps `$reduceMethod $in into $out attr-dict type($in) into`type($out)`

Or `$out = $reduceMethod ( $in )`

Reply:
Yes, I could change that -- I actually did muse over this and was torn between the two.
I thought I should actually explain why I went this format.  It was mainly because in the tensor format, it would appear as: 

$reduceMethod $in into $out attr-dict type($in) intotype($out)` --> type($result) which I found less clearer compared to standard DPS style with optional returns (which what the proposed is). However, the proposed is inconsistent with other rock ops fancier assembly format -- which Im not sure that Im a fan of -- I do prefer simpler (input operands, output dps operands) : op input types --> op output types.

I think I would vote for consistency and move to your suggestion now but should explain for the record.

Reply:
Entirely fair on the DPS situation. I wrote the `rock.*` assembly formats on the principle that they mostly appear in `memref` form.

And, even with DPS, you still get `out`/`return` redundancies.

Reply:
Done!